lib/libXcomposite/man/Xcomposite.man
'\" t
.\"     Title: Xcomposite
.\"    Author: Keith Packard <keithp@keithp.com>
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: <pubdate>23 April 2007</pubdate>
.\"    Manual: X Composite Extension Library
.\"    Source: __vendorversion__
.\"  Language: English
.\"
.TH "XCOMPOSITE" "__libmansuffix__" "<pubdate>23 April 2007</pubdate>" "__vendorversion__" "X Composite Extension Library"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
Xcomposite \- X Composite Extension library
.SH "SYNOPSIS"
.sp
.ft B
.nf
#include <X11/extensions/Xcomposite\&.h>
.fi
.ft
.HP \w'Bool\ XCompositeQueryExtension('u
.BI "Bool XCompositeQueryExtension(Display\ *" "dpy" ", int\ *" "event_base_return" ", int\ *" "error_base_return" ");"
.HP \w'Status\ XCompositeQueryVersion('u
.BI "Status XCompositeQueryVersion(Display\ *" "dpy" ", int\ *" "major_version_return" ", int\ *" "minor_version_return" ");"
.HP \w'int\ XCompositeVersion('u
.BI "int XCompositeVersion(void);"
.HP \w'void\ XCompositeRedirectWindow('u
.BI "void XCompositeRedirectWindow(Display\ *" "dpy" ", Window\ " "window" ", int\ " "update" ");"
.HP \w'void\ XCompositeRedirectSubwindows('u
.BI "void XCompositeRedirectSubwindows(Display\ *" "dpy" ", Window\ " "window" ", int\ " "update" ");"
.HP \w'void\ XCompositeUnredirectWindow('u
.BI "void XCompositeUnredirectWindow(Display\ *" "dpy" ", Window\ " "window" ", int\ " "update" ");"
.HP \w'void\ XCompositeUnredirectSubwindows('u
.BI "void XCompositeUnredirectSubwindows(Display\ *" "dpy" ", Window\ " "window" ", int\ " "update" ");"
.HP \w'XserverRegion\ XCompositeCreateRegionFromBorderClip('u
.BI "XserverRegion XCompositeCreateRegionFromBorderClip(Display\ *" "dpy" ", Window\ " "window" ");"
.HP \w'Pixmap\ XCompositeNameWindowPixmap('u
.BI "Pixmap XCompositeNameWindowPixmap(Display\ *" "dpy" ", Window\ " "window" ");"
.HP \w'Window\ XCompositeGetOverlayWindow('u
.BI "Window XCompositeGetOverlayWindow(Display\ *" "dpy" ", Window\ " "window" ");"
.HP \w'void\ XCompositeReleaseOverlayWindow('u
.BI "void XCompositeReleaseOverlayWindow(Display\ *" "dpy" ", Window\ " "window" ");"
.SH "DESCRIPTION"
.PP
The composite extension provides several related mechanisms:
.PP
Per\-hierarchy storage
.RS 4
The rendering of an entire hierarchy of windows is redirected to off\-screen storage\&. The pixels of that hierarchy are available whenever it is viewable\&. Storage is automatically reallocated when the top level window changes size\&. Contents beyond the geometry of the top window are not preserved\&.
.RE
.PP
Automatic shadow update
.RS 4
When a hierarchy is rendered off\-screen, the X server provides an automatic mechanism for presenting those contents within the parent window\&. The implementation is free to make this update lag behind actual rendering operations by an unspecified amount of time\&. This automatic update mechanism may be disabled so that the parent window contents can be completely determined by an external application\&.
.RE
.PP
Composite Overlay Window
.RS 4
Version 0\&.3 of the protocol adds the Composite Overlay Window, which provides compositing managers with a surface on which to draw without interference\&. This window is always above normal windows and is always below the screen saver window\&. It is an InputOutput window whose width and height are the screen dimensions\&. Its visual is the root visual and its border width is zero\&. Attempts to redirect it using the composite extension are ignored\&. This window does not appear in the reply of the QueryTree request\&. It is also an override redirect window\&. These last two features make it invisible to window managers and other X11 clients\&. The only way to access the XID of this window is via the CompositeGetOverlayWindow request\&. Initially, the Composite Overlay Window is unmapped\&.
.RE
.PP
Parent window clipping
.RS 4
Version 0\&.4 of the protocol modifies the semantics of parent window clipping in the presence of manual redirected children\&. With this version, the area in the parent covered by manual redirected children is left in the parent clip list instead of being removed as in older versions\&.
.RE
.PP
Per\-hierarchy storage may be created for individual windows or for all children of a window\&. Manual shadow update may be selected by only a single application for each window; manual update may also be selected on a per\-window basis or for each child of a window\&. Detecting when to update may be done with the Damage extension\&.
.PP
The off\-screen storage includes the window contents, its borders and the contents of all descendants\&.
.SH "ARGUMENTS"
.PP
.PP
\fIdisplay\fR
.RS 4
Pointer to the
\fBDisplay\fR
structure returned from
\fBXOpenDisplay\fR
for the connection to the X server\&.
.RE
.PP
\fIevent_base_return\fR
.RS 4
Pointer to integer where the base value for Composite Extension events will be stored\&.
.RE
.PP
\fIerror_base_return\fR
.RS 4
Pointer to integer where the base value for Composite Extension errors will be stored\&.
.RE
.PP
\fImajor_version_return\fR
.RS 4
Pointer to integer where the major version of the Composite Extension supported by the X server will be stored\&.
.RE
.PP
\fIminor_version_return\fR
.RS 4
Pointer to integer where the minor version of the Composite Extension supported by the X server will be stored\&.
.RE
.PP
\fIwindow\fR
.RS 4
Specifies the window ID to operate on\&.
.RE
.PP
\fIupdate\fR
.RS 4
Specifies the mode for updating the window contents\&. Must be either
\fBCompositeRedirectAutomatic\fR
or
\fBCompositeRedirectManual\fR\&.
.RE
.SH "FUNCTIONS"
.PP
.PP
\fBXCompositeQueryExtension\fR
.RS 4

\fBXCompositeQueryExtension\fR
determines if the Composite Extension is available on the given display\&. It returns
\fBTrue\fR
if the extension is supported, otherwise
\fBFalse\fR\&. If the extension is present, the base values for events and errors are returned, and can be used to decode incoming event and error values\&.
.RE
.PP
\fBXCompositeQueryVersion\fR
.RS 4

\fBXCompositeQueryVersion\fR
determines if the X Server supports a version of the X Composite Extension which is compatible with the client library\&. A non\-zero Status is returned if a compatible version of the extension is supported, otherwise a zero Status is returned\&. If the extension is supported, the major and minor version numbers are returned to indicate the level of Composite Extension support\&. No other XComposite functions (except XCompositeQueryExtension) may be called before this function\&. If a client violates this rule, the effects of all subsequent XComposite calls that it makes are undefined\&.
.RE
.PP
\fBXCompositeVersion\fR
.RS 4

\fBXCompositeVersion\fR
returns the version of the X Composite library\&. The version number is encoded as:
(major * 10000) + (minor * 100) + revision
.sp
For example, version 1\&.4\&.6 would be encoded as the integer 10406\&.
.RE
.PP
\fBXCompositeRedirectWindow\fR
.RS 4

\fBXCompositeRedirectWindow\fR
requests the X server to direct the hierarchy starting at
\fIwindow\fR
to off\-screen storage\&. The
\fIupdate\fR
argument specifies whether the contents are mirrored to the parent window automatically or not\&. Only one client at a time may specify an update type of
\fBCompositeRedirectManual\fR, another attempt will result in a
BadAccess
error\&. When all clients enabling redirection terminate, the redirection will automatically be disabled\&.
.sp
The root window may not be redirected\&. Doing so results in a
BadMatch
error\&. Specifying an invalid window id will result in a
BadWindow
error\&.
.RE
.PP
\fBXCompositeRedirectSubwindows\fR
.RS 4

\fBXCompositeRedirectSubwindows\fR
requests the X server to redirect hierarchies starting at all current and future children of
\fIwindow\fR
as in
\fBXCompositeRedirectWindow\fR\&. If
\fIupdate\fR
is
\fBCompositeRedirectManual\fR, then painting of the window background during window manipulation and ClearArea requests is inhibited\&.
.RE
.PP
\fBXCompositeUnredirectWindow\fR
.RS 4

\fBXCompositeUnredirectWindow\fR
requests the X server to terminate redirection of
\fIwindow\fR\&. If the specified window was not selected for redirection by the current client, a
BadValue
error results\&.
.RE
.PP
\fBXCompositeUnredirectSubwindows\fR
.RS 4

\fBXCompositeUnredirectWindow\fR
requests the X server to terminate redirection of all children of
\fIwindow\fR\&. If the specified window was not selected for sub\-redirection by the current client, a
BadValue
error results\&.
.RE
.PP
\fBXCompositeCreateRegionFromBorderClip\fR
.RS 4

\fBXCompositeCreateRegionFromBorderClip\fR
creates a region containing the "usual" border clip value; that is the area of the window clipped against siblings and the parent\&. This region can be used to restrict rendering to suitable areas while updating only a single window\&. The region is copied at the moment the request is executed; future changes to the window hierarchy will not be reflected in this region\&.
.RE
.PP
\fBXCompositeNameWindowPixmap\fR
.RS 4

\fBXCompositeNameWindowPixmap\fR
creates and returns a pixmap id that serves as a reference to the off\-screen storage for
\fIwindow\fR\&. This pixmap will remain allocated until freed, even if the window is unmapped, reconfigured or destroyed\&. However, the window will get a new pixmap allocated each time it is mapped or resized, so this function will need to be reinvoked for the client to continue to refer to the storage holding the current window contents\&. Generates a
BadMatch
error if
\fIwindow\fR
is not redirected or is not visible\&.
.sp
The X server must support at least version 0\&.2 of the Composite Extension for
\fBXCompositeNameWindowPixmap\fR\&.
.RE
.PP
\fBXCompositeGetOverlayWindow\fR
.RS 4

\fBXCompositeGetOverlayWindow\fR
returns the window ID of the Composite Overlay Window for the screen specified by the argument
\fIwindow\fR\&. This function notifies the X server that the client wishes to use the Composite Overlay Window of this screen\&. If this Composite Overlay Window has not yet been mapped, it is mapped by this request\&.
.sp
The Composite Overlay Window for a particular screen will be unmapped when all clients who have called this function have either called
\fBXCompositeReleaseOverlayWindow\fR
for that screen, or terminated their connection to the X server\&.
.sp
The X server must support at least version 0\&.3 of the Composite Extension for
\fBXCompositeGetOverlayWindow\fR\&.
.RE
.PP
\fBXCompositeReleaseOverlayWindow\fR
.RS 4
This request specifies that the client is no longer using the Composite Overlay Window on the screen specified by the argument
\fIwindow\fR\&. A screen\'s Composite Overlay Window is unmapped when there are no longer any clients using it\&.
.sp
The X server must support at least version 0\&.3 of the Composite Extension for
\fBXCompositeReleaseOverlayWindow\fR\&.
.RE
.SH "AUTHORS"
.PP
\fBKeith Packard\fR <\&keithp@keithp\&.com\&>
.RS 4
Extension specification and implementation
.RE
.PP
\fBDeron Johnson\fR <\&deron\&.johnson@sun\&.com\&>
.RS 4
Overlay Window specification and implementation
.RE


xserver/ChangeLog
commit e597f0119cd69b6d9edf86d06d941468f90d8e6d
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 27 21:07:17 2012 -0800

    configure.ac: 1.11.4
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit 6c491463ee7a3637c8aae7ddae9625b36b792078
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 20 23:54:58 2012 -0800

    1.11.3.902 (1.11.4 RC2)
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit b747a91485184649c05aba6fa10734f70e926032
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Wed Jan 18 11:47:21 2012 -0800

    XQuartz: Bump bundle version to 2.7.1
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit a1ed67fa0551ee2b39d002d3b786a2017b37c7ac
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Fri Jan 6 13:20:45 2012 +1000

    render: don't bother with animated cursors on floating slaves (#39989)
    
    X.Org Bug 39989 <http://bugs.freedesktop.org/show_bug.cgi?id=39989>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit bbb6b8c834e0e1491ca14403b5d0840dd14380d3)

commit f89c39e6c9e20ffc3c1ae228ee601c2ab9007ff4
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Mon Jan 9 15:04:10 2012 +1000

    dix: Update pointer limits for floating devices too (#43635)
    
    When the screen is restructured, the pointer limits need to be reset for
    floating slave devices as well, not just for master pointers. Only skip
    devices that don't have a cursor (attached slaves and keyboard)
    
    Bug reproducer: float an absolute slave device, rotate the screen - the
    device is now confined to a section of the screen only.
    
    X.Org Bug 43635 <http://bugs.freedesktop.org/show_bug.cgi?id=43635>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Chase Douglas <chase.douglas@canonical.com>
    (cherry picked from commit 35bd77e9d0701daae87d681900d749604fc6471f)

commit dea9377a257a402b2c7fe4a96c958d951a7d4579
Author: Zhigang Gong <zhigang.gong@linux.intel.com>
Date:   Wed Jan 4 07:01:19 2012 +0000

    mi/mibitblt: Fix an overflow bug of bit shift.
    
    When depth equal to 32 and planeMask equal to 0, the overflow will
    occur and cause the pixmap can't be cleared. There are some test
    cases in XTS hit this bug, and this fix can eliminate the corresponding
    failures.
    
    Signed-off-by: Zhigang Gong <zhigang.gong@linux.intel.com>
    Reviewed-by: Adam Jackson <ajax@redhat.com>
    Signed-off-by: Keith Packard <keithp@keithp.com>
    (cherry picked from commit f4956faab9ccf9aba6cf9603f4489f5dad19a347)

commit d2db5100b1bafeeea8c1febd390631c8f8abb094
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Fri Oct 28 21:29:50 2011 -0700

    Convert DetermineClientCmd to use strdup instead of malloc+strncpy
    
    *cmdname is initialized to NULL earlier in the function, so it's
    okay to overwrite it with NULL if strdup fails, don't need that
    extra check.
    
    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit 780133f9ae7fada462714b47e79d26075bbd9abe)

commit 5b28bcb340e279e7f4e76617e86b439092367690
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:23:12 2011 -0500

    dix: Tune dixLookupDrawable for success
    
    The vast vast vast majority of resource lookups are successful.  Move some
    work to the error paths so we don't punish success.
    
    Before:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1148346.9/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2091666.1/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 8f9bdfd293ad8e45755efe8d764b4dcc2a724f51)

commit e1ea93fec93b13d18df4451210aaaf322b769b6a
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 20:23:40 2011 -0500

    dix: Tune dtrace hooks around Dispatch
    
    Don't call LookupMajorName if the hooks aren't active, it's quite expensive.
    
    Before:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 83a98543b58c661a22574a6f8d8f9d777c0955b8)

commit 9eaa0b5c89f645c50ea90e07e8b49ebacf4349a0
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:00:05 2011 -0500

    xace: ricer tuning for XaceHook
    
    gcc doesn't want to hoist the check for XaceHooks[hook] != NULL above the
    varargs code for some reason, so do it ourselves.
    
    Before:
    40000000 trep @   0.0010 msec (1050420.2/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (1921147.6/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 11977fab546da35d579ebe285e3c26864007805e)

commit a3b6ecc893e17028cfbe16d8d7981c9037fc0766
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Tue Jan 3 10:45:45 2012 +1000

    xfree86: split warning about missing identifier or input driver
    
    Check for identifier first and bail if it's missing (also remove the current

lib/libXcomposite/src/Xcomposite.c
/*
 * $Id: Xcomposite.c,v 1.5 2019/06/19 19:29:23 matthieu Exp $
 *
 * Copyright © 2006 Sun Microsystems
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of Sun Microsystems not be used in
 * advertising or publicity pertaining to distribution of the software without
 * specific, written prior permission.  Sun Microsystems makes no
 * representations about the suitability of this software for any purpose.  It
 * is provided "as is" without express or implied warranty.
 *
 * SUN MICROSYSTEMS DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO
 * EVENT SHALL SUN MICROSYSTEMS BE LIABLE FOR ANY SPECIAL, INDIRECT OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE,
 * DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
 * TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THIS SOFTWARE.
 *
 * Copyright © 2003 Keith Packard
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of Keith Packard not be used in
 * advertising or publicity pertaining to distribution of the software without
 * specific, written prior permission.  Keith Packard makes no
 * representations about the suitability of this software for any purpose.  It
 * is provided "as is" without express or implied warranty.
 *
 * KEITH PACKARD DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO
 * EVENT SHALL KEITH PACKARD BE LIABLE FOR ANY SPECIAL, INDIRECT OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE,
 * DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
 * TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THIS SOFTWARE.
 */

#include "xcompositeint.h"

XCompositeExtInfo XCompositeExtensionInfo;

const char XCompositeExtensionName[] = COMPOSITE_NAME;

/*
 * XCompositeExtRemoveDisplay - remove the indicated display from the
 * extension object. (Replaces XextRemoveDisplay.)
 */
static int
XCompositeExtRemoveDisplay (XCompositeExtInfo *extinfo, Display *dpy)
{
    XCompositeExtDisplayInfo *info, *prev;

    /*
     * locate this display and its back link so that it can be removed
     */
    _XLockMutex(_Xglobal_lock);
    prev = NULL;
    for (info = extinfo->head; info; info = info->next) {
	if (info->display == dpy) break;
	prev = info;
    }
    if (!info) {
	_XUnlockMutex(_Xglobal_lock);
	return 0;		/* hmm, actually an error */
    }

    /*
     * remove the display from the list; handles going to zero
     */
    if (prev)
	prev->next = info->next;
    else
	extinfo->head = info->next;

    extinfo->ndisplays--;
    if (info == extinfo->cur) extinfo->cur = NULL;  /* flush cache */
    _XUnlockMutex(_Xglobal_lock);

    Xfree ((char *) info);
    return 1;
}

static int
XCompositeCloseDisplay (Display *dpy, XExtCodes *codes)
{
    return XCompositeExtRemoveDisplay (&XCompositeExtensionInfo, dpy);
}

/*
 * XCompositeExtAddDisplay - add a display to this extension. (Replaces
 * XextAddDisplay)
 */
static XCompositeExtDisplayInfo *
XCompositeExtAddDisplay (XCompositeExtInfo	*extinfo,
			 Display		*dpy,
			 const char		*ext_name)
{
    XCompositeExtDisplayInfo    *info;

    info = (XCompositeExtDisplayInfo *) Xmalloc (sizeof (XCompositeExtDisplayInfo));
    if (!info) return NULL;
    info->display = dpy;

    info->codes = XInitExtension (dpy, ext_name);

    /*
     * if the server has the extension, then we can initialize the
     * appropriate function vectors
     */
    if (info->codes) {
	xCompositeQueryVersionReply	rep;
	xCompositeQueryVersionReq	*req;
        XESetCloseDisplay (dpy, info->codes->extension,
                           XCompositeCloseDisplay);
	/*
	 * Get the version info
	 */
	LockDisplay (dpy);
	GetReq (CompositeQueryVersion, req);
	req->reqType = info->codes->major_opcode;
	req->compositeReqType = X_CompositeQueryVersion;
	req->majorVersion = COMPOSITE_MAJOR;
	req->minorVersion = COMPOSITE_MINOR;
	if (!_XReply (dpy, (xReply *) &rep, 0, xTrue))
	{
	    UnlockDisplay (dpy);
	    SyncHandle ();
	    Xfree(info);
	    return NULL;
	}
	info->major_version = rep.majorVersion;
	info->minor_version = rep.minorVersion;
	UnlockDisplay (dpy);
	SyncHandle ();
    } else {
	/* The server doesn't have this extension.
	 * Use a private Xlib-internal extension to hang the close_display
	 * hook on so that the "cache" (extinfo->cur) is properly cleaned.
	 * (XBUG 7955)
	 */
	XExtCodes *codes = XAddExtension(dpy);
	if (!codes) {
	    Xfree(info);
	    return NULL;
	}
        XESetCloseDisplay (dpy, codes->extension, XCompositeCloseDisplay);
    }

    /*
     * now, chain it onto the list
     */
    _XLockMutex(_Xglobal_lock);
    info->next = extinfo->head;
    extinfo->head = info;
    extinfo->cur = info;
    extinfo->ndisplays++;
    _XUnlockMutex(_Xglobal_lock);
    return info;
}

/*
 * XCompositeExtFindDisplay - look for a display in this extension; keeps a
 * cache of the most-recently used for efficiency. (Replaces
 * XextFindDisplay.)
 */
static XCompositeExtDisplayInfo *
XCompositeExtFindDisplay (XCompositeExtInfo *extinfo,
		      Display	    *dpy)
{
    XCompositeExtDisplayInfo *info;

    /*
     * see if this was the most recently accessed display
     */
    if ((info = extinfo->cur) && info->display == dpy)
	return info;

    /*
     * look for display in list
     */
    _XLockMutex(_Xglobal_lock);
    for (info = extinfo->head; info; info = info->next) {
	if (info->display == dpy) {
	    extinfo->cur = info;     /* cache most recently used */
	    _XUnlockMutex(_Xglobal_lock);
	    return info;
	}
    }
    _XUnlockMutex(_Xglobal_lock);

    return NULL;
}

XCompositeExtDisplayInfo *
XCompositeFindDisplay (Display *dpy)
{
    XCompositeExtDisplayInfo *info;

    info = XCompositeExtFindDisplay (&XCompositeExtensionInfo, dpy);
    if (!info)
	info = XCompositeExtAddDisplay (&XCompositeExtensionInfo, dpy,
				    XCompositeExtensionName);
    return info;
}


Bool
XCompositeQueryExtension (Display *dpy,
			  int *event_base_return,
			  int *error_base_return)
{
    XCompositeExtDisplayInfo *info = XCompositeFindDisplay (dpy);

    if (XCompositeHasExtension(info))
    {
	*event_base_return = info->codes->first_event;
	*error_base_return = info->codes->first_error;
	return True;
    }
    else
	return False;
}

Status
XCompositeQueryVersion (Display *dpy,
		    int	    *major_version_return,
		    int	    *minor_version_return)
{
    XCompositeExtDisplayInfo	*info = XCompositeFindDisplay (dpy);

    XCompositeCheckExtension (dpy, info, 0);
    *major_version_return = info->major_version;
    *minor_version_return = info->minor_version;
    return 1;
}

int
XCompositeVersion (void)
{
    return XCOMPOSITE_VERSION;
}

void
XCompositeRedirectWindow (Display *dpy, Window window, int update)
{
    XCompositeExtDisplayInfo	    *info = XCompositeFindDisplay (dpy);
    xCompositeRedirectWindowReq	    *req;

    XCompositeSimpleCheckExtension (dpy, info);
    LockDisplay (dpy);
    GetReq (CompositeRedirectWindow, req);
    req->reqType = info->codes->major_opcode;
    req->compositeReqType = X_CompositeRedirectWindow;
    req->window = window;
    req->update = update;
    UnlockDisplay (dpy);
    SyncHandle ();
}

void
XCompositeRedirectSubwindows (Display *dpy, Window window, int update)
{
    XCompositeExtDisplayInfo	    *info = XCompositeFindDisplay (dpy);
    xCompositeRedirectSubwindowsReq *req;

    XCompositeSimpleCheckExtension (dpy, info);
    LockDisplay (dpy);
    GetReq (CompositeRedirectSubwindows, req);
    req->reqType = info->codes->major_opcode;
    req->compositeReqType = X_CompositeRedirectSubwindows;
    req->window = window;
    req->update = update;
    UnlockDisplay (dpy);
    SyncHandle ();
}

void
XCompositeUnredirectWindow (Display *dpy, Window window, int update)
{
    XCompositeExtDisplayInfo	    *info = XCompositeFindDisplay (dpy);
    xCompositeUnredirectWindowReq   *req;

    XCompositeSimpleCheckExtension (dpy, info);
    LockDisplay (dpy);
    GetReq (CompositeUnredirectWindow, req);
    req->reqType = info->codes->major_opcode;
    req->compositeReqType = X_CompositeUnredirectWindow;
    req->window = window;
    req->update = update;
    UnlockDisplay (dpy);
    SyncHandle ();
}

void
XCompositeUnredirectSubwindows (Display *dpy, Window window, int update)
{
    XCompositeExtDisplayInfo		*info = XCompositeFindDisplay (dpy);
    xCompositeUnredirectSubwindowsReq	*req;

    XCompositeSimpleCheckExtension (dpy, info);
    LockDisplay (dpy);
    GetReq (CompositeUnredirectSubwindows, req);
    req->reqType = info->codes->major_opcode;
    req->compositeReqType = X_CompositeUnredirectSubwindows;
    req->window = window;
    req->update = update;
    UnlockDisplay (dpy);
    SyncHandle ();
}

XserverRegion
XCompositeCreateRegionFromBorderClip (Display *dpy, Window window)
{
    XCompositeExtDisplayInfo		    *info = XCompositeFindDisplay (dpy);
    xCompositeCreateRegionFromBorderClipReq *req;
    XserverRegion			    region;

    XCompositeCheckExtension (dpy, info, 0);
    LockDisplay (dpy);
    GetReq (CompositeCreateRegionFromBorderClip, req);
    req->reqType = info->codes->major_opcode;
    req->compositeReqType = X_CompositeCreateRegionFromBorderClip;
    req->window = window;
    region = req->region = XAllocID (dpy);
    UnlockDisplay (dpy);
    SyncHandle ();
    return region;
}

Pixmap
XCompositeNameWindowPixmap (Display *dpy, Window window)
{
    XCompositeExtDisplayInfo	    *info = XCompositeFindDisplay (dpy);
    xCompositeNameWindowPixmapReq   *req;
    Pixmap			    pixmap;

    XCompositeCheckExtension (dpy, info, 0);
    LockDisplay (dpy);
    GetReq (CompositeNameWindowPixmap, req);
    req->reqType = info->codes->major_opcode;
    req->compositeReqType = X_CompositeNameWindowPixmap;
    req->window = window;
    pixmap = req->pixmap = XAllocID (dpy);
    UnlockDisplay (dpy);
    SyncHandle ();
    return pixmap;
}

Window
XCompositeGetOverlayWindow (Display *dpy, Window window)
{
    XCompositeExtDisplayInfo	    *info = XCompositeFindDisplay (dpy);
    xCompositeGetOverlayWindowReq   *req;
    xCompositeGetOverlayWindowReply rep;

    XCompositeCheckExtension (dpy, info, 0);
    LockDisplay (dpy);
    GetReq (CompositeGetOverlayWindow, req);
    req->reqType = info->codes->major_opcode;
    req->compositeReqType = X_CompositeGetOverlayWindow;
    req->window = window;
    if (!_XReply (dpy, (xReply *) &rep, 0, xFalse))
    {
	UnlockDisplay (dpy);
	SyncHandle ();
	return 0;
    }

    UnlockDisplay (dpy);
    SyncHandle ();

    return rep.overlayWin;
}

void
XCompositeReleaseOverlayWindow (Display *dpy, Window window)
{
    XCompositeExtDisplayInfo	    *info = XCompositeFindDisplay (dpy);
    xCompositeReleaseOverlayWindowReq   *req;

    XCompositeSimpleCheckExtension (dpy, info);
    LockDisplay (dpy);
    GetReq (CompositeReleaseOverlayWindow, req);
    req->reqType = info->codes->major_opcode;
    req->compositeReqType = X_CompositeReleaseOverlayWindow;
    req->window = window;
    UnlockDisplay (dpy);
    SyncHandle ();
}
